services/jvm/src/main/scala/io/gearpump/services/MasterService.scala

```diff
@@ -33,194 +36,185 @@ import io.gearpump.cluster.worker.WorkerSummary
-                  complete(write(
-                    MasterService.Status(success=false, reason="Jar file not supplied")))
-                }
-                val userConf = MasterService.findFormDataOption(fields, "conf")
-                val argsArray = args.getOrElse("").split(" +")
-
-                onComplete(Future(
-                  MasterService.submitJar(jar.get, userConf, argsArray, system.settings.config))) {
-                  case Success(_) =>
+        } ~
+          path("applist") {
+            onComplete(askActor[AppMastersData](master, AppMastersDataRequest)) {
+              case Success(value: AppMastersData) =>
+                complete(write(value))
+              case Failure(ex) => failWith(ex)
+            }
+          } ~
+          path("workerlist") {
+            def future = askActor[WorkerList](master, GetAllWorkers).flatMap { workerList =>
+              val workers = workerList.workers
+              val workerDataList = List.empty[WorkerSummary]
+
+              Future.fold(workers.map { workerId =>
+                askWorker[WorkerData](master, workerId, GetWorkerData(workerId))
+              })(workerDataList) { (workerDataList, workerData) =>
+                workerDataList :+ workerData.workerDescription
+              }
+            }
+            onComplete(future) {
+              case Success(result: List[WorkerSummary]) => complete(write(result))
+              case Failure(ex) => failWith(ex)
+            }
+          } ~
+          path("config") {
+            onComplete(askActor[MasterConfig](master, QueryMasterConfig)) {
+              case Success(value: MasterConfig) =>
+                val config = Option(value.config).map(_.root.render()).getOrElse("{}")
+                complete(config)
+              case Failure(ex) =>
+                failWith(ex)
+            }
+          } ~
+          path("metrics" / RestPath) { path =>
+            parameter("readLatest" ? "false") { readLatestInput =>
+              val readLatest = Try(readLatestInput.toBoolean).getOrElse(false)
+              val query = QueryHistoryMetrics(path.head.toString, readLatest)
+              onComplete(askActor[HistoryMetrics](master, query)) {
+                case Success(value) =>
+                  complete(write(value))
+                case Failure(ex) =>
+                  failWith(ex)
+              }
+            }
+          } ~
+          path("submitapp") {
+            post {
+              parameters("args") { args =>
+                uploadFile { fileMap =>
+                  val jar = fileMap.get("jar").map(_.file)
+                  val userConf = fileMap.get("conf").map(_.file)
+
+                  if (jar.isEmpty) {
                     complete(write(
-                      MasterService.Status(success=true)))
-                  case Failure(ex) =>
-                    failWith(ex)
+                      MasterService.Status(success = false, reason = "Jar file not supplied")))
```

Comment:
I'd recommend `MasterService.Status` to be `MasterService.Response` or `RestService.Response`.